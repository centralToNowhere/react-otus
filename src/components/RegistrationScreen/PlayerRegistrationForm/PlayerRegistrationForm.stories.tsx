import React from "react";
import { Story, Meta } from "@storybook/react";
import {
  PlayerRegistrationForm,
  IRegistrationFormProps,
} from "@/components/RegistrationScreen/PlayerRegistrationForm";

export default {
  title: "Components/RegistrationScreen/PlayerRegistrationForm",
  component: PlayerRegistrationForm,
  args: {
    player: {
      registered: true,
      name: "Player1",
    },
    loginPending: false,
    loginError: "",
    onPlayerRegistration: () => {
      //empty
    },
  },
} as Meta;

const Template: Story<IRegistrationFormProps> = (args) => {
  return <PlayerRegistrationForm {...args} />;
};

export const PlayerRegistrationFormDefault = Template;

export const PlayerRegistrationFormPending = Template.bind({});
PlayerRegistrationFormPending.args = {
  loginPending: true,
};

export const PlayerRegistrationFormError = Template.bind({});
PlayerRegistrationFormError.args = {
  loginError: "Player is not registered!",
};
