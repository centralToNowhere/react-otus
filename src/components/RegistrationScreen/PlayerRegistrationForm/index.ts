export { PlayerRegistrationForm } from "./PlayerRegistrationForm";
export type { IRegistrationFormProps } from "./PlayerRegistrationForm";
