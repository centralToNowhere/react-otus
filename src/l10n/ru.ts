export const l10n = {
  maxWidthLabel: "Макс. ширина:",
  maxHeightLabel: "Макс. высота:",
  cellSizeLabel: "Размер ячейки:",
  capacityLabel: "Процент заполненности:",
  speedLabel: "Обновлений в секунду:",
  buttonStart: "Старт",
  buttonStop: "Стоп",
  buttonReset: "Сброс",
  buttonGenerateRandom: "Создать",

  // Экран регистрации
  gameNameFull: "Conway's Game of Life",
  gameHeadingPart1: "Conway's",
  gameHeadingPart2: "Game of Life",
  buttonStartGameAsPlayer: "Начать игру",
  rulesHeading: "Правила",
  description1: "Игра «Жизнь» (англ. ",
  source: "Источник:",
  sourceName: "Игра «Жизнь»",
  description2:
    ") — клеточный автомат, придуманный английским математиком Джоном Конвеем в 1970 году.",
  rule1:
    "Место действия этой игры — «вселенная» — это размеченная на клетки поверхность или плоскость — безграничная, ограниченная, или замкнутая (в пределе — бесконечная плоскость).",
  rule2:
    "Каждая клетка на этой поверхности может находиться в двух состояниях: быть «живой» (заполненной) или быть «мёртвой» (пустой). Клетка имеет восемь соседей, окружающих её.",
  rule3:
    "Распределение живых клеток в начале игры называется первым поколением. Каждое следующее поколение рассчитывается на основе предыдущего по таким правилам:",
  rule3_1:
    "в пустой (мёртвой) клетке, рядом с которой ровно три живые клетки, зарождается жизнь",
  rule3_2:
    "если у живой клетки есть две или три живые соседки, то эта клетка продолжает жить; в противном случае, если соседей меньше двух или больше трёх, клетка умирает («от одиночества» или «от перенаселённости»)",
  rule4: "Игра прекращается, если:",
  rule4_1: "на поле не останется ни одной «живой» клетки",
  rule4_2:
    "конфигурация на очередном шаге в точности (без сдвигов и поворотов) повторит себя же на одном из более ранних шагов (складывается периодическая конфигурация)",
  rule4_3:
    "при очередном шаге ни одна из клеток не меняет своего состояния (складывается стабильная конфигурация; предыдущее правило, вырожденное до одного шага назад)",

  // Контейнер игрока
  registerPlayerLabel: "Введите имя игрока",
  logoutButton: "Выйти",
  generation: "Поколение",

  // Ошибки
  minCellsAmount: "Размер игрового поля меньше размера ячейки.",
  maxCellsAmount: "Превышено максимальное количество ячеек",
  positiveNumber: "Ожидается положительное число.",
  nonNegativeNumber: "Ожидается неотрицательное число.",
};
