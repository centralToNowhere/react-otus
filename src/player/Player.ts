import { IFieldControlState } from "@/components/Fields";
import { IAuthState } from "@/auth";

export interface IPlayer {
  registered: boolean;
  name: string | null;
}

export const defaultPlayer: IPlayer = {
  registered: false,
  name: null,
};

export type PlayerStorageData = IFieldControlState &
  Omit<IAuthState, "loginPending">;
