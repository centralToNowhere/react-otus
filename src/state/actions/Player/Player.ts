export interface IPlayer {
  registered: boolean,
  name: string | null
}

export const isPlayer = (player: IPlayer): player is IPlayer => {
  return player && typeof player === "object"
    && typeof player.registered === "boolean"
    && (typeof player.name === "string" || player.name === null);
}