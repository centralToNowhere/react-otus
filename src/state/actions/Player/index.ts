export * from "./Player";
export * from "./SetPlayerAction";
