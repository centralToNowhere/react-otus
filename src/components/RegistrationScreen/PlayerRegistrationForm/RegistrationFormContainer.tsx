import React, { FC } from "react";
import { PlayerRegistrationForm } from "@/components/RegistrationScreen/PlayerRegistrationForm/index";
import { useSelector } from "react-redux";
import { useRegistration, selectLoginPending, selectLoginError } from "@/auth";

export const RegistrationFormContainer: FC = () => {
  const [player, onPlayerRegistration] = useRegistration();
  const loginPending = useSelector(selectLoginPending);
  const loginError = useSelector(selectLoginError);
  const onPlayerRegName = (playerName: string | null) => {
    onPlayerRegistration(playerName);
  };

  return (
    <PlayerRegistrationForm
      onPlayerRegistration={onPlayerRegName}
      player={player}
      loginPending={loginPending}
      loginError={loginError}
    />
  );
};
