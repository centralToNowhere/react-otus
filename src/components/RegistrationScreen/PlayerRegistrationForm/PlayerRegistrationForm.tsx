import React, { FC, useCallback, useState, SyntheticEvent } from "react";
import { l10n } from "@/l10n/ru";
import { IPlayer } from "@/state/actions";
import { FormField } from "@/components/Form/FormField";
import { FormElement } from "@/components/Form";
import { css } from "@emotion/react";
import { InputField, LabelField } from "@/components/Fields";
import { FormButton } from "@/components/Buttons";

export interface IRegistrationFormProps {
  player: IPlayer;
  onPlayerRegistration: (playerName: string | null) => void;
}

export const PlayerRegistrationForm: FC<IRegistrationFormProps> = (props) => {
  const [playerName, setPlayerName] = useState(props.player.name);

  const onPlayerNameChange = useCallback(
    (e: SyntheticEvent<HTMLInputElement>) => {
      setPlayerName(e.currentTarget.value);
    },
    [setPlayerName]
  );

  const registerPlayer = () => {
    props.onPlayerRegistration(playerName);
  };

  return (
    <FormElement
      css={css`
        background: #e8d5f0;
        width: 70%;
        max-width: 600px;
        display: inline-block;
      `}
    >
      <FormField
        css={css`
          display: flex;
        `}
      >
        <div
          css={css`
            margin-bottom: 50px;
          `}
        >
          <LabelField htmlFor="register-player-input">
            {l10n.registerPlayerLabel}
          </LabelField>
          <InputField
            css={css`
              width: 100%;
            `}
            id={"register-player-input"}
            value={playerName || ""}
            onChange={onPlayerNameChange}
            name="playerName"
            autoFocus={true}
          />
        </div>
        <FormButton
          type="button"
          css={css`
            margin-right: auto !important;
          `}
          name="startGameAsPlayer"
          onClick={registerPlayer}
        >
          {l10n.buttonStartGameAsPlayer}
        </FormButton>
      </FormField>
    </FormElement>
  );
};
