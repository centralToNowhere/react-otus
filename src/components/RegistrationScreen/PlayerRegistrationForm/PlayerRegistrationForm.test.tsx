import React from "react";
import { render, screen, waitFor } from "@testing-library/react";
import { PlayerRegistrationForm } from "@/components/RegistrationScreen/PlayerRegistrationForm/PlayerRegistrationForm";
import { l10n } from "@/l10n/ru";
import userEvent from "@testing-library/user-event/dist";
import {defaultPlayer} from "@/player/Player";

describe("PlayerRegistrationForm tests", () => {
  it("should render player registration form", () => {
    const { asFragment } = render(
      <PlayerRegistrationForm
        player={defaultPlayer}
        onPlayerRegistration={jest.fn()}
      />
    );

    const input = screen.getByLabelText(l10n.registerPlayerLabel);
    const button = screen.getByRole("button");

    expect(input).toBeInTheDocument();
    expect(button).toBeInTheDocument();
    expect(asFragment()).toMatchSnapshot();
  });

  it("should call onPlayerRegistration on button click", async () => {
    const onPlayerRegistration = jest.fn();

    render(
      <PlayerRegistrationForm
        player={defaultPlayer}
        onPlayerRegistration={onPlayerRegistration}
      />
    );

    const button = screen.getByRole("button");

    userEvent.click(button);

    await waitFor(() => {
      expect(onPlayerRegistration).toHaveBeenCalledTimes(1);
    });
  });
});
