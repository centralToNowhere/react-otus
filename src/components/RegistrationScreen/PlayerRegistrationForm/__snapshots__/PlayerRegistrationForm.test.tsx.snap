// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PlayerRegistrationForm tests should render player registration form 1`] = `
<DocumentFragment>
  <form
    class="css-ldog09-FormElement-PlayerRegistrationForm e1dt6zfi0"
  >
    <div
      class="css-kmw9t5-FormField-PlayerRegistrationForm eybdsu10"
    >
      <div
        class="css-gj9mjj-PlayerRegistrationForm"
      >
        <label
          class="css-m0dhjx-LabelField e1gymppe0"
          for="register-player-input"
        >
          Введите имя игрока
        </label>
        <input
          class="css-gtg0cp-InputField-PlayerRegistrationForm e15roiqv0"
          id="register-player-input"
          name="playerName"
          value=""
        />
      </div>
      <button
        class="css-ssn7vi-FormButton-PlayerRegistrationForm e2n0d7s0"
        name="startGameAsPlayer"
        type="button"
      >
        Начать игру
      </button>
    </div>
    <div
      class="css-1a4u2lo-SpinnerContainer eok4aqp0"
    >
      <div
        class="css-rsmjug-PlayerRegistrationForm spinner-border"
        role="status"
      />
    </div>
  </form>
</DocumentFragment>
`;
